solver/jobs.go

```diff
@@ -593,21 +595,40 @@ func (b *dumbBuilder) build(ctx context.Context, e Edge) (CachedResult, error) {
 
 		cm, err := edge.op.CacheMap(ctx, int(e.Index))
 		if err != nil {
+			mu.Unlock()
```

Comment:
Do we still need the lock at L551? I was originally locking everything in the loop as a stop-gap. 

I think it would be clearer to move the lock/unlock calls directly before and after `execInProgress` & `cache` read/writes.

Reply:
I was able to move the initial lock from 551 down to 591 (just before the for loop where we start checking if `cache` has an entry).

We still need to hang on to that lock up until we make an entry in execInProgress for the particular digest; in order to prevent other jobs from starting up a second exec on the same digest.

Reply:
Should we move the lock & `execInProgress` to the top of the loop near L555?

Reply:
so pretty much back to where it was :laughing: 

(but after the first return on error which was a deadlock)

I'll move it up there.

Reply:
Nah, I mean locking the `execInProgress` check & setting independently at the top of the loop rather than locking over some of the non-critical-section code.